File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    public int solution(String begin, String target, String[] words) {
        if (!find(target, words))
            return 0;
        
        int n = words.length;
        String[] newWords = new String[n + 1];
        
        for (int i = 0; i < n; i++)
            newWords[i+1] = words[i];
        newWords[0] = begin;
                
        HashMap<String, Integer> getIdx = new HashMap<>();
        for (int i = 0; i <= n; i++)
            getIdx.put(newWords[i], i);
        
        List<List<Integer>> graph = new ArrayList<>();
        for (int unused = 0; unused <= n; unused++)
            graph.add(new ArrayList<>());
        for (int i = 0; i <= n; i++) {
            for (int j = i+1; j <= n; j++) {
                if (compare(newWords[i], newWords[j])) {
                    graph.get(i).add(j);
                    graph.get(j).add(i);
                }
            }
        }
            
        return bfs(graph, n+1, 0, getIdx.get(target));
    }
    
    public boolean find(String word, String[] words) {
        for (String w: words) {
            if (word.equals(w))
                return true;
        }
        return false;
    }
    
    public boolean compare(String s1, String s2) {
        int count = 0;
        for (int idx = 0; idx < s1.length(); idx++) {
            if (s1.charAt(idx) != s2.charAt(idx))
                count++;
        }
        return count == 1 ? true : false;
    }
    
    public int bfs(List<List<Integer>> graph, int n, int start, int last) {
        boolean[] visited = new boolean[n];
        visited[start] = true;
        
        ArrayDeque<int[]> q = new ArrayDeque<>();
        q.add(new int[]{start, 0});
        
        while (!q.isEmpty()) {
            int[] first = q.poll();
            int now = first[0];
            int depth = first[1];
            
            if (now == last)
                return depth;
            
            for (int nxt: graph.get(now)) {
                if (visited[nxt])
                    continue;
                
                visited[nxt] = true;
                q.add(new int[]{nxt, depth+1});
            }
        }
        
        return 0;
    }
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def solution(begin, target, words):
    if target not in words:
        return 0
    
    words.append(begin)
    idx = len(words[0])
    n = len(words)
    graph = [[] for _ in range(n)]
    for i, wi in enumerate(words):
        for j, wj in enumerate(words[:i]):
            if sum(map(lambda i: 1 if wi[i] != wj[i] else 0, range(idx))) == 1:
                graph[i].append(j)
                graph[j].append(i)
                
    return bfs(graph, n, n-1, words.index(target))
    

def bfs(graph, n, start, destination):
    visited = [0] * n
    visited[start] = 1
    
    q = deque()
    q.append((start, 0))
    
    while q:
        now, count = q.popleft()
        if now == destination:
            return count
        
        for nxt in graph[now]:
            if not visited[nxt]:
                visited[nxt] = 1
                q.append((nxt, count+1))
                
    return 0            
    